core/lib/zksync_core/src/basic_witness_input_producer/vm_interactions.rs

```diff
@@ -0,0 +1,130 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::state_keeper::io::common::load_l1_batch_params;
+
+use multivm::VmInstance;
+use vm::{HistoryEnabled, L2BlockEnv};
+use zksync_config::constants::{
+    SYSTEM_CONTEXT_ADDRESS, SYSTEM_CONTEXT_CURRENT_L2_BLOCK_INFO_POSITION,
+};
+use zksync_dal::StorageProcessor;
+use zksync_state::{PostgresStorage, PostgresStorageCaches, ReadStorage, StorageView};
+use zksync_types::block::unpack_block_info;
+use zksync_types::{
+    AccountTreeId, L1BatchNumber, L2ChainId, MiniblockNumber, StorageKey, Transaction,
+};
+use zksync_utils::h256_to_u256;
+
+use tokio::runtime::Handle;
+
+pub(super) async fn create_vm<'a>(
+    rt_handle: Handle,
+    l1_batch_number: L1BatchNumber,
+    mut connection: StorageProcessor<'a>,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+) -> (
+    VmInstance<PostgresStorage<'a>, HistoryEnabled>,
+    Rc<RefCell<StorageView<PostgresStorage<'a>>>>,
+) {
+    let prev_l1_batch_number = L1BatchNumber(l1_batch_number.0 - 1);
+    let (_, miniblock_number) = connection
+        .blocks_dal()
+        .get_miniblock_range_of_l1_batch(prev_l1_batch_number)
+        .await
+        .unwrap()
+        .unwrap_or_else(|| {
+            panic!(
+                "l1_batch_number {:?} must have a previous miniblock to start from",
+                l1_batch_number
+            )
+        });
+
+    let fee_account_addr = connection
+        .blocks_dal()
+        .get_fee_address_for_l1_batch(&l1_batch_number)
+        .await
+        .unwrap()
+        .unwrap_or_else(|| {
+            panic!(
+                "l1_batch_number {:?} must have fee_address_account",
+                l1_batch_number
+            )
+        });
+    let (system_env, l1_batch_env) = load_l1_batch_params(
+        &mut connection,
+        l1_batch_number,
+        fee_account_addr,
+        validation_computational_gas_limit,
+        l2_chain_id,
+    )
+    .await
+    .unwrap();
+
+    let pg_storage = PostgresStorage::new(rt_handle.clone(), connection, miniblock_number, true)
+        .with_caches(PostgresStorageCaches::new(
+            128 * 1_024 * 1_024, // 128MB -- picked as the default value used in EN
```

Comment:
it's only valuable if we want to cache it between  l1 batches. Is it the case here?

Reply:
...but then the cache must be created externally, and not for each VM instance. Plus we must expect that the batches will be processed in order, otherwise cache will be mostly invalidated (e.g. caches won't really work if there will be multiple producers running).